### ISSUE TITLE
Make FakeScope threadsafe

### ISSUE THREAD

**Issue by TimLovellSmith** [FakeItEasy/FakeItEasy#484]
![whattt](https://cloud.githubusercontent.com/assets/2711795/7458929/43b123e4-f24d-11e4-89ea-54eeb42efc71.png)

This is really confusing me. How can this be a failure - doesn't that mean the 1 call it got matches?
The code in question looks like: 

```C#
var fakeDns = A.Fake<IAzureDnsResourceManager>();

/// do stuff

var theCall = A.CallTo(() => fakeDns.GetCNameAync("busy123"));
theCall.Returns(Task.FromResult("busy123456"));

/// do stuff

theCall.MustHaveHappened();
```

FakeItEasy version 1.25.1. What I'm doing *different* from before is running tests in parallel... if I rerun this test on its own, it passes.

**Comment by TimLovellSmith** [comment]
Ah, looks like this is issue #60. Sorry :)

**Comment by blairconrad** [comment]
@TimLovellSmith, I completely believe that you're running afoul of #60, but are you sure that the code you're having problems with is

`var theCall = A.CallTo(() => fakeDns.GetCNameAync("busy123"));`?

I have only ever seen problems when we use `That.Matches` or `Ignored`, and I'd be very excited/sad to find out that the multiple threads are causing problems with other kinds of verifications.

(Also, your screenshoot appears to include a `Ignored`, so that makes me feel hopeful, but I thought I'd check just in case.)

Thanks.

**Comment by blairconrad** [comment]
Oh, and if you're interested, you may want to take the nupkg from
http://teamcity.codebetter.com/viewLog.html?buildId=189740&tab=artifacts&buildTypeId=bt929#!132is1d

for a spin. It contains my proposed fix for #478, which is a facet of #60.

**Comment by TimLovellSmith** [comment]
Yes, I'm completely sure that's the code I'm having problem with. I'll give your nupkg a spin and see if it helps, thanks!

**Comment by TimLovellSmith** [comment]
So. I'm seeing less problems with the new nupkg, I think. But I'm still seeing some problems, and the problem basically seems to be that the fake call that *should* be configured did not actually get invoked. Here is some code to explain. I have this line of code:

StorageAndAffinityGroup sa = await AzureResourceManager.CreateStorageAccountInAnySubscription(scope, Pool, dep.Name, dep.Region, suggestedName, token);

That call is supposed to get faked out by the following:

var storageCall = A.CallTo(() => worker.AzureResourceManager.CreateStorageInAffinityGroup(null, null, null, null, null, A<CancellationToken>._)).WithAnyArguments();
            storageCall.Returns(Task.FromResult(new StorageAccount() { ConnectionString = "DefaultEndpointsProtocol=https;AccountName=storagesample;AccountKey=KWPLd0rpW2T0U7K2pVpF8rYr1BgYtR7wYQk33AYiXeUoquiaY6o0TWqduxmPHlqeCNZ3LU0DHptbeIHy5l/Yhg==" }));

What actually happens though, after going multithreaded, is that sometimes, CreateStorageAccountInAnySubscription returns a storage account with a connection string that is null, and the test fails with null reference exception in the code under test, on a line like this:

return CloudStorageAccount.Parse(this.ConnectionString).Credentials.AccountName;

(throws NullReferenceException: Value cannot be null. Parameter name: connectionString)

Could this be happening because there are some collections holding the fake call configurations to be executed that are shared across fake objects and not thread-safe?

**Comment by TimLovellSmith** [comment]
Maybe you can tell from my question - I am suspicious that the problem is FakeManager is not thread-safe.

**Comment by TimLovellSmith** [comment]
reopened/renamed issue.

**Comment by adamralph** [comment]
I reformatted the comments to use [fenced code blocks](https://help.github.com/articles/github-flavored-markdown/#fenced-code-blocks).

**Comment by blairconrad** [comment]
@TimLovellSmith, thanks for the update. It may be that the `FakeManager` is not threadsafe, but there is one FakeManager per fake, so unless the fakes are shared between threads, I'd be surprised if that's what was causing the multi-threaded problem. Then again, it's been a little while since I looked at that specifically. I can take another peek soon.
In the meantime (and I'm not trying to get you to _all_ my work for me, just most), do you have a slightly more complete example that shows the problem? Seeing the plumbing around your tests may shed some additional light on the problem.

Thanks.

**Comment by TimLovellSmith** [comment]
So, this particular test is an async Task rather than void, and uses owin TestServer to 'send' a payload to a web api controller and receive a response. There's lots of asyncness, but I think there is no asyncness and only one thread while the test setup code and test assertion code is running.

In between the test setup and the test assertion, the code is very likely running in another thread/(s). Not that that means FakeManager is definitely the problem...

 Here is the whole test, excluding some of the helper/product code:

```C#
        [Fact]
        public async Task NameAvailabilityReturns409ForOccupiedName()
        {
            var fakeDns = A.Fake<IAzureDnsResourceManager>();
            var controller = FakeHelper.Create<NamesController>(
                new Tuple<Type, object>(typeof(IAzureDnsResourceManager), fakeDns));
            var builder = FakeHelper.GetContainerBuilder(false).Register(controller);

            var theCall = A.CallTo(() => fakeDns.GetCNameAync("busy123"));
            theCall.Returns(Task.FromResult("busy123456"));
            
            using (var server = new SubclassedTestServer(builder.Build()))
            {
                var uri = CreateUri(action: "checkNameAvailability");
                var request = CreateRequest(server, uri, "resourceType", "busy123");
                var response = await request.PostAsync();
                Assert.Equal(409, (int)response.StatusCode);
            }

            theCall.MustHaveHappened();
        }
```

**Comment by TimLovellSmith** [comment]
I think I figured out what makes the tests to fail..
I added some code like this in FakeScope.Create
public static IFakeScope Create(IFakeObjectContainer container)

            var result = new ChildScope(Current, container);
            Current = result;
            throw new Exception("blow up the world!");

And now basically the same set of tests fails with Exception("blow up the world!").

The problem is that we are using FakeScope.Create()! 

Result Message:	System.Exception : blow up the world!
Result StackTrace:	
at FakeItEasy.Core.FakeScope.Create(IFakeObjectContainer container) in c:\repo\FakeItEasy\Source\FakeItEasy\Core\FakeScope.cs:line 48
   at FakeItEasy.Core.FakeScopeFactory.Create(IFakeObjectContainer container) in c:\repo\FakeItEasy\Source\FakeItEasy\Core\FakeScopeFactory.cs:line 13
   at FakeItEasy.FakeFacade.CreateScope(IFakeObjectContainer container) in c:\repo\FakeItEasy\Source\FakeItEasy\FakeFacade.cs:line 35
   at FakeItEasy.Fake.CreateScope(IFakeObjectContainer container) in c:\repo\FakeItEasy\Source\FakeItEasy\Fake.cs:line 51
   at AzCacheRP.Tests.FakeHelper.Create[T](Tuple`2[] values) in

**Comment by blairconrad** [comment]
I edited your comment to improve the code/output formatting.

Ah. That was the key. Secretly using `Fake.CreateScope`. If only we'd had the helper code all along.
Yeah, I'd suspected that scopes would mess up parallel test execution, but they aren't documented and so I assume are infrequently used and I can't think of a way to fix them, so I hadn't jumped on the problem, in favour of concentrating on work for 2.0, as well as outside stuff I have going on.
Thanks, @TimLovellSmith.

**Comment by blairconrad** [comment]
The unthreadsafety of the scopes also caused problems in #528.

**Comment by IharBury** [comment]
I see two possible ways of dealing ways of dealing with this issue:
1) CreateScope method can be marked obsolete with clarification that it doesn't support parallel tests
2) FakeScope.Current can be modified to store the value in ExecutionContext instead of static field. This will make all fakes a bit slower but will allow child contexts in parallel tests including async code support.

**Comment by adamralph** [comment]
> 1) CreateScope method can be marked obsolete with clarification that it doesn't support parallel tests

If we decide to throw in the towel then we can just remove it from 2.0, rather than marking obsolete, since that is a major breaking release anyway.

**Comment by blairconrad** [comment]
>  > 1) CreateScope method can be marked obsolete with clarification that it doesn't support parallel tests

> If we decide to throw in the towel then we can just remove it from 2.0, rather than marking obsolete, since that is a major breaking release anyway.

It's true. I'm not a scope fan, personally. I think they have limited use, and would prefer to encourage other patterns, such as just throwing away fakes and getting new ones or reconfiguring methods.
On that front, I wouldn't mind removing them. Also I think that removing `CreateScope(IFakeObjectContainer )` would allow us to clean up some stuff around `IFakeObjectContainer` that I was afraid to in #374.

However, scopes are essential to performing [Ordered Assertions](https://github.com/FakeItEasy/FakeItEasy/wiki/Ordered-assertions#details), as currently implemented, and I fear that people would feel the loss.
And at least @matkoch's TestFx uses `CreateScope` [as part of the framework](https://github.com/matkoch/TestFx/search?utf8=%E2%9C%93&q=CreateScope) (but I'm not sure why).

So I'm a little torn.

In any event, I wouldn't advocate removing them solely because they're not threadsafe. If somebody really loves the scopes and would have their tests broken by parallel test execution as in XUnit 2.x's default configuration, they may be happy just reducing concurrency to 1 and keeping the scopes. Happier than they would be if they woke up with no scopes one day.

So, I think the first task is to establish whether we want to remove scopes from 2.0.

My take: from a selfish maintainer point of view, I'd love it. But ultimately, I don't think it's the right move for our clients, at least not without providing an alternative for the ordered assertions.

**Comment by matkoch** [comment]
@blairconrad, I appreciate that you thought of me :grinning:

Of course I'd like to argue to keep the API. I've been a very intensive FIE user and I'm very ambitious about testing. Based on my experience, I'd say that using fake scopes should mostly be avoided, but there are cases, when you really need them. So taking them away from v2 would really be a pain, and also may be a reason to switch to another framework or fork FIE... I don't know.

Speaking about a solution to *warn* users, why don't you just implement that method thread safe? Create a mutex or something like that?!

**Comment by adamralph** [comment]
Agreed that fixing it is the most desirable outcome.

I propose we write a failing test and spike fixing it before making further assumptions.

**Comment by blairconrad** [comment]
@matkoch, thanks for chiming in.

I also think fixing is best.

Off the top of my head, there are two things to make sure work:

1. creating a Scope in one thread should not adversely affect Scopes from other threads. Certainly we've seen situations where two threads are running and the CurrentScope is changed underneath one of them.
2. when we spawn a new thread (logical or otherwise) we should be (initially) participate in the CurrentScope that the spawning thread is.

**Comment by IharBury** [comment]
I've created a pull request for this: https://github.com/FakeItEasy/FakeItEasy/pull/575. But it breaks FakeScopeSpecs.CreatingFakeInsideScope scenario. Any ideas how to fix it?

**Comment by adamralph** [comment]
An alternative to this, which avoids all the problems inherit in ambient state, is to dump the static scope and make it instance based instead. E.g.

```C#
using (var scope = Fake.CreateScope())
{
    fake.Bar(1);
    fake.Bar(2);
    using (scope.OrderedAssertions())
    {
        A.CallTo(() => fake.Bar(2)).MustHaveHappened();
        exception = Record.Exception(() => A.CallTo(() => fake.Bar(1)).MustHaveHappened());
    }
}
```

would become:

```C#
using (var scope = Fake.CreateScope())
{
    fake.In(scope).Bar(1);
    fake.In(scope).Bar(2);
    using (var assertions = scope.OrderedAssertions())
    {
        A.CallTo(() => fake.Bar(2)).In(scope).MustHaveHappened();
        exception = Record.Exception(() => A.CallTo(() => fake.Bar(1)).In(scope).MustHaveHappened());
    }
}
```

**Comment by blairconrad** [comment]
@adamralph, in your example, nothing ensures that `fake.Bar(1)` is considered to have happened on the right scope.
That's where the real ambient state magic happens.

**Comment by adamralph** [comment]
@blairconrad I also chained a `InScope(scope)` onto the `A.CallTo()`.

I've no idea if what I'm suggesting is even possible though...

**Comment by adamralph** [comment]
... or if the API would be considered acceptable...

**Comment by blairconrad** [comment]
@adamralph, I missed the change to `fake.In(scope).Bar(1);`, but even so, most of the time the call to `Bar` will be made by _production code_, which should have no concept of Fakes, let along Fake Scopes.

**Comment by adamralph** [comment]
@blairconrad a call to `fake.Bar(1)` must be getting recorded somehow into the ambient state for things like ordered assertions. In my proposal, we'd make this state explicit instead of ambient.

**Comment by blairconrad** [comment]
I fear I am sorely misunderstanding you, @adamralph (and I thought I was no longer febrile), but we _can't_. The call to `fake.Bar(1)` will be made by _production code_. It will never know that any sort of Current Fake Scope State exists, let alone which state that is.

**Comment by adamralph** [comment]
@blairconrad ok, I see what you mean now. In a real scenario the (production code) SUT won't know to call `fake.In(scope).Bar(1);` rather than `fake.Bar(1);`.

So the crux of the problem is somehow making the fake aware of a specific scope rather than it having to pull out the ambient scope.

How about something like:

```C#
var dep = A.Fake<IDep>();
var sut = new Sut(dep);

using (var scope = Fake.CreateScope())
{
    dep.SetScope(scope);

    sut.Foo():

    using (scope.OrderedAssertions())
    {
        A.CallTo(() => fake.Bar()).MustHaveHappened();
        A.CallTo(() => fake.Baz()).MustHaveHappened();
    }
}
```

We could also have overloads for `Fake.CreateScope()` which accept a set of fakes and calls the `SetScope()` method on them, and/or an overload of `A.Fake<T>()` which accepts a scope and does the same.

This might all be crazy, I'm just throwing ideas out there...

**Comment by blairconrad** [comment]
`or an overload of A.Fake<T>() which accepts a scope and does the same.` would be required, I'd think, considering that (today) we use the current Scope in order to help construct the fakes.
Although that _was_ a usage I was arguing against recently. 

Also, in order to fully support the current range of Scope operations, we'd need to be able to set and unset scopes from fakes at will also, similar to the methods you suggest.

Oh, and be aware that `OrderedAssertions()` sets shared state - a static member on the class, so the ball of  yarn we're untangling is getting larger…
I'm not trying to discourage you, just to properly frame the size of the problem.

(I think the most useful part of the current FakeScope class is that is allows ordered assertions, and I wonder if we wouldn't be able to support that _without_ Scopes. If so, I'd look again a the possibility of ditching the Scopes. The other Scope functionality _might_ be useful, but I'm a little doubtful.)

**Comment by thomaslevesque** [comment]
I like the idea of an explicit scope rather than an ambient one, if it's possible, but it seems to be far from trivial.

> I think the most useful part of the current FakeScope class is that is allows ordered assertions, and I wonder if we wouldn't be able to support that without Scopes

I also think it would be preferable. Using scopes for ordered assertions isn't intuitive at all. I would prefer something like this:

    Calls.To(
        () => x.Foo(42),
        () => y.Bar(),
        () => x.Baz())
        .MustHaveHappenedInOrder();

Is there any other compelling use case for fake scopes?

If they're not thread-safe an cannot be made thread-safe, I'd be in favor of dropping the feature completely...

**Comment by IharBury** [comment]
@thomaslevesque, I like your proposed syntax. And it can be easily implemented by adding a statically unique timestamp to each recorded call (IFakeObjectCall). The timestamp can be made thread-safe and quick via Interlocked.Increment.

**Comment by blairconrad** [comment]
> Is there any other compelling use case for fake scopes?

For me, no. Other than checking the order of the calls to a fake, you can

1. make calls to a Fake, enrol in a Scope, make a few more calls and then assert _only on the calls in the scope_. I think some people may like that because it allows you to disregard calls made during setup. Me, I've never thought it much use. And…
2. create a Scope that uses a custom `IFakeObjectContainer`, but I doubt anyone does this, and I've suggested getting rid of it before…

Then again, I'm just one guy. What do I know? @matkoch probably uses fake scopes more than anyone, in [TestFx](https://github.com/matkoch/TestFx), but I can't tell what benefit they give there, and even he doesn't plug in custom `IFakeObjectContainer`s.

**Comment by blairconrad** [comment]
@IharBury, I'm intrigued by your suggested implementation. As it is, every call is added to a master list anyhow, which enforces an ordering, and I'd thought that would be sufficient. But it's not threadsafe. Then again, if you have to protect against concurrent calls anyhow, asserting that your calls happened in a particular order will probably not end well…

**Comment by IharBury** [comment]
@blairconrad, I haven't explored the storage of the calls in detail. I would expect calls being stored in different collections per each fake object, when using the root scope. So that the order of calls to different fake objects cannot be determined from the storages themselves.

Anyway that is already thread-safe, when each fake object is only used from one thread, as I understand.

**Comment by matkoch** [comment]
@blairconrad, @IharBury 
Suggested API looks okay. I was only again obsoleting ordered assertions at all... the concrete usage is more or less unimportant.

**Comment by thomaslevesque** [comment]
> As it is, every call is added to a master list anyhow, which enforces an ordering, and I'd thought that would be sufficient.

Yes, but this list is per-fake, right? So we still need a way to tell if a call to `x` happened before or after a call to `y`.

> But it's not threadsafe.

It could probably be made thread-safe without too much hassle (by just taking a lock when accessing the list)

> Then again, if you have to protect against concurrent calls anyhow, asserting that your calls happened in a particular order will probably not end well…

Agreed

**Comment by blairconrad** [comment]
@thomaslevesque, I was thinking of the per-(scope, fake) list, not the per-fake list. More on that in one sentence!

@IharBury, well spotted. I didn't notice that `RootScope.OnAddRule` has an empty implementation. I thought all Scopes tracked their calls.

> It could probably be made thread-safe without too much hassle (by just taking a lock when accessing the list)

Absolutely, but if we want to just keep a per-fake list of calls and not involve scopes, I think @IharBury's suggestion may be preferable.

**Comment by blairconrad** [comment]
@matkoch, insulted? Why? No insult was intended! You are our "star scope user".
Oh, I'm sure there are private projects using Scopes, but one can only reason about what one sees.

**Comment by matkoch** [comment]
That's why I argue that removing the feature completely is a bad idea... My only objection =)

**Comment by adamralph** [comment]
I like where this discussion is heading... (ditching scopes)

Any proposed syntax for ordered assertions in a single statement would also need to cater for all the various options that can be applied to each call assertion, e.g. `Repeated.Twice`, etc.

**Comment by thomaslevesque** [comment]
> Any proposed syntax for ordered assertions in a single statement would also need to cater for all the various options that can be applied to each call assertion, e.g. Repeated.Twice, etc.

Good point. We could do something like this:

```
Calls.Ordered(a => 
{
    a.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once);
    a.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice);
    a.CallTo(() => x.Baz()).MustHaveHappened();
});
```

but it's a bit verbose IMO...

**Comment by thomaslevesque** [comment]
Another idea, more concise:

```
Ordered.Calls(to => 
{
    to(() => x.Foo(42)).Once();
    to(() => y.Bar()).Twice();
    to(() => x.Baz());
}).MustHaveHappened();
```

**Comment by blairconrad** [comment]
1. Should we move discussion to another issue, specifically one around decoupling ordered assertions from scopes (and then attack the remainder of scopes later)? 
2. @thomaslevesque, I didn't find your example in https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-184751019 to be overly verbose, or rather I feel that mirroring the non-ordered scope provides more benefits than changing the syntax to remove a few characters does.

**Comment by adamralph** [comment]
@thomaslevesque that's not bad.

What about another approach:

```C#
A
    .CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once)
    .AndThen
    .A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice)
    .AndThen
    .A.CallTo(() => y.Baz()).MustHaveHappened();
```

Mind you, I don't really like the necessity to chain, since then this can't be spread across xbehave steps, so I'd prefer to be able to hang the assertions off some object, e.g.

```C#
"Then x must have been foo'ed with 42 once"
    .x(() =>
    {
        a = new OrderedAssertions();
        a.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once);
    });

"And y must have been barred twice"
    .x(() => a.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice));

"Then y must have been bazzed"
    .x(() => a.CallTo(() => x.Baz()).MustHaveHappened());
```

**Comment by adamralph** [comment]
@blairconrad sure, let's move the ordered assertions discussion.

**Comment by adamralph** [comment]
So it looks like we've killed off one reason to keep `FakeScope` by agreeing on a [proposal](https://github.com/FakeItEasy/FakeItEasy/issues/602) to introduce an instance based ordered assertions feature. What other features is `FakeScope` serving?

**Comment by blairconrad** [comment]
Off the top of my head, I know no more than I said in https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-184679735:

> Other than checking the order of the calls to a fake, you can
> 1. make calls to a Fake, enrol in a Scope, make a few more calls and then assert _only on the calls in the scope_. I think some people may like that because it allows you to disregard calls made during setup. Me, I've never thought it much use. And…
> 2. create a Scope that uses a custom `IFakeObjectContainer`, but I doubt anyone does this, and I've suggested getting rid of it before…

For point 2, the new `IDummyFactory` and `IFakeOptionsBuilder`s should more than compensate for the loss.

There may also be other uses, but I can't check now (work, and worse, a new computer today), and am out all evening…

**Comment by IharBury** [comment]
@blairconrad, for point 1, I think, it's also possible to use timestamps instead of storing the calls in a scope. We just need the scope to remember the initial timestamp and the final timestamp during the scope lifecycle.

**Comment by adamralph** [comment]
@IharBury what benefit does your suggestion offer? Does it allow the removal of ambient state? 

Generally though, I'm hesitant to use timestamps for anything like this since their uniqueness is non-deterministic unless you start enforcing delays between calls, which may have a substantial effect over a large test suite.

**Comment by adamralph** [comment]
@blairconrad I'm wondering if we can just drop those two features... they seem of little practical use and neither NSubstitute or Moq seem to have equivalent features. The alternative, of course, is to make scope instance based and come up with some way to pass the scope to the various methods, as we are doing with the proposed assertions object in #602

**Comment by IharBury** [comment]
@adamralph, I was referring to timestamps not as datetime values but as (`long`) integer values. They can and generated thread-safely and quick via Interlocked.Increment of a static field. Each `IFakeObjectCall` can store its timestamp.

In that case scopes do not need to be associated with calls at all. They can just record the current timestamp on creation and on disposal resulting in a timestamp range. All the calls with timestamps in that range can be considered belonging to that scope.

**Comment by adamralph** [comment]
What if two tests are running in parallel?

**Comment by blairconrad** [comment]
@IharBury: 
> @blairconrad, for point 1, I think, it's also possible to use timestamps instead of storing the calls in a scope. We just need the scope to remember the initial timestamp and the final timestamp during the scope lifecycle.

Yup, you mentioned. I thought it had merit. Although either keeping a master list of calls across objects or the timestamps would seem to work about as well.

@adamralph: 
> what benefit does [@IharBury]'s  suggestion offer? Does it allow the removal of ambient state?

It's a way to keep track of the order of calls made to various fakes. Currently, the `ChildScope` keeps a list of all calls made within the scope.

@adamralph:
> Generally though, I'm hesitant to use timestamps for anything like this since their uniqueness is non-deterministic unless you start enforcing delays between calls, which may have a substantial effect over a large test suite.

@IharBury:
>  I was referring to timestamps not as datetime values but as (long) integer values. They can and generated thread-safely and quick via Interlocked.Increment of a static field.

@adamralph 
> What if two tests are running in parallel?

Well, one of the calls has to be first, so one call will get a lower timestamp. Had we been using a "master list", one would've been inserted first.
The `MustHaveHappened` will know on which fake the call was made, so there shouldn't be any opportunity for confusion, _unless someone is sharing Fakes between tests_, which they oughtn't do.

**Comment by adamralph** [comment]
Perhaps I'm missing something, but still I don't see how the timestamp proposal removes the need for ambient state.

If I make a call to `MustHaveHappened`, how does it now within which scope to make the assertion, regardless of whether timestamps are used or not?

**Comment by blairconrad** [comment]
It wouldn't know at the `MustHaveHappened` point. In the latest version of the syntax we had (I think there's room for tweaking):

```c#
var callOrder = new CallOrderAsserter(); // actual name to be defined
A.CallTo(() => x.Foo(42)).MustHaveHappened(Repeated.Once).In(callOrder);
A.CallTo(() => y.Bar()).MustHaveHappened(Repeated.Twice).In(callOrder);
A.CallTo(() => x.Baz()).MustHaveHappened().In(callOrder);
```

We'd do the in-order checking in `In`. It's the `In` that makes us know we need to check order.
In my head, the `callOrder` would be initialized with a "before the beginning of time" value for the "last call I matched", and as we walk down the `In` in the sample code, if we find the matching calls, then the `callOrder` would increase its "last call I matched" value to match the highest of the calls that it matched. That way we'd make sure that we'd only consider matching (for the sake of order) the `y.Bar` calls with a timestamp higher than the timestamp of the last `x.Foo(42)` that we matched.

**Comment by blairconrad** [comment]
… so no scope. We match against all calls made on `x` and `y`, but every time `callOrder` is used, it "ratchets" the timestamp, so we never look back at a call that happened earlier than the last one we matched.

**Comment by adamralph** [comment]
OK, so, as I understand it, the timestamps are simplified way of implementing instance based scoped assertions or ordered assertions....

Note that the [original suggestion](https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-185558833) from @IharBury was only with reference to the scoped assertions point in [the reply](https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-185322100) to my query about remaining uses of scope but I can see how it can apply to ordered assertions too.

**Comment by adamralph** [comment]
...not that we yet have a syntax proposal for instance based scoped assertions...

**Comment by blairconrad** [comment]
> Note that the original suggestion from @IharBury was only with reference to the scoped assertions point in the reply to my query about remaining uses of scope but I can see how it can apply to ordered assertions too.

Ah, yes. I misread that entirely. Me, I'm liking the immediate application to ordered assertions. If it turns out that people clamour for more scopey applications, we could investigate using the timestamps for those purposes as well.

**Comment by blairconrad** [comment]
> ...not that we yet have a syntax proposal for instance based scoped assertions...

No. But we're getting close?

**Comment by adamralph** [comment]
For ordered assertions, yes, but for scoped assertions?

**Comment by adamralph** [comment]
I'm not sure it's even possible for scoped assertions, since we're back to the problem of the production code making the calls to the fakes, so they can't be associated with any kind of "scope" instance.

**Comment by adamralph** [comment]
Incidentally, NSub has a [`ClearReceivedCalls()` method](http://nsubstitute.github.io/help/clear-received-calls/) which would seem to satisfy the use mentioned above for scoped assertions.

**Comment by blairconrad** [comment]
> For ordered assertions, yes, but for scoped assertions?

I need to read better. Yes.

> I'm not sure it's even possible for scoped assertions, since we're back to the problem of the production code making the calls to the fakes, so they can't be associated with any kind of "scope" instance.

Well, that's always been the case. If we consider creating a Scope as a "range of fake call timestamps", we might be able to get back the scoped assertions functionality that we (soon hope to) used to have.

> Incidentally, NSub has a ClearReceivedCalls() method which would seem to satisfy the use case mentioned above for scoped assertions.

Yeah, I'd figured we could always go that kind of route if we need to. I'm half-inclined to just take the functionality away during 2.0 and then add it back if people feel a need. I think all these use cases are kind of weird, to be brutally honest. But not everyone works the same way, with the same objects.

**Comment by adamralph** [comment]
> I'm half-inclined to just take the functionality away during 2.0 and then add it back if people feel a need.

I'm :+1: for that. It will also drive out the actual use cases so we can provide more specific solutions, e.g. an equivalent of `ClearReceivedCalls()`.

**Comment by TimLovellSmith** [comment]
> I'm half-inclined to just take the functionality away during 2.0 and then add it back if people feel a need.
Seems like an amazingly smart and practical idea to me!

**Comment by blairconrad** [comment]
> Seems like an amazingly smart and practical idea to me!

You made my day _again_, @TimLovellSmith. How much to have you follow me around all the time, making comments like this?

**Comment by TimLovellSmith** [comment]
(btw, reopened by accident - closed is fine by me!)

**Comment by adamralph** [comment]
I didn't touch it, so I guess someone moved it on the Waffle board...

Anyway, closed is good.

Replaced by #602, #603 and #604.

**Comment by DanielRose** [comment]
Well, that change bit me. The reason is stated in [this comment](https://github.com/FakeItEasy/FakeItEasy/issues/484#issuecomment-184679735):

> 1. make calls to a Fake, enrol in a Scope, make a few more calls and then assert only on the calls in the scope. I think some people may like that because it allows you to disregard calls made during setup.

I use this in a few places for some class-level fakes where there may be spurious calls which I want to disregard, especially in combination with `A.CallTo().MustNotHaveHappened()`.

The change means I need to recreate those fakes for every test (which I can do of course). An alternative would be the previously mentioned "ClearAllCalls()".

**Comment by blairconrad** [comment]
> The change means I need to recreate those fakes for every test (which I can do of course)

That's what I'd recommend in any case.

**Comment by adamralph** [comment]
@DanielRose :+1: to what @blairconrad said. Reusing fake instances between various methods is usually a path to pain. If you don't want to repeat set up code, you can create some helper methods to create your fakes, which ultimately can become a DSL for your tests.
